src/script/lua_api/l_craft.cpp

```diff
@@ -326,118 +326,87 @@ int ModApiCraft::l_get_craft_result(lua_State *L)
 	return 2;
 }
 
-// get_craft_recipe(result item)
-int ModApiCraft::l_get_craft_recipe(lua_State *L)
+int get_craft_recipe_helper(lua_State *L, IGameDef *gdef, bool return_multiple)
 {
-	NO_MAP_LOCK_REQUIRED;
-
-	int k = 1;
-	int input_i = 1;
-	std::string o_item = luaL_checkstring(L,input_i);
+	std::string o_item = luaL_checkstring(L, 1);
 
-	IGameDef *gdef = getServer(L);
 	ICraftDefManager *cdef = gdef->cdef();
 	CraftInput input;
-	CraftOutput output(o_item,0);
-	bool got = cdef->getCraftRecipe(input, output, gdef);
-	lua_newtable(L); // output table
-	if(got){
-		lua_newtable(L);
-		for(std::vector<ItemStack>::const_iterator
-			i = input.items.begin();
-			i != input.items.end(); i++, k++)
-		{
-			if (i->empty())
-			{
+	CraftOutput output(o_item, 0);
+
+	std::vector<CraftDefinition*> recipes_list = cdef->getCraftRecipes(output, gdef);
+
+	lua_createtable(L, return_multiple ? recipes_list.size() : 0, 0);
+
+	if (recipes_list.empty()) {
+		if (return_multiple) {
+			lua_pushnil(L);
+		} else {
+			lua_pushnil(L);
+			lua_setfield(L, -2, "items");
+			setintfield(L, -1, "width", 0);
+		}
+		return 1;
+	}
+
+	std::vector<CraftDefinition*>::const_iterator iter = recipes_list.begin();
+	for (u16 i = 0; iter != recipes_list.end(); iter++) {
+		CraftOutput tmpout;
+		tmpout.item = "";
+		tmpout.time = 0;
+		tmpout = (*iter)->getOutput(input, gdef);
+		input = (*iter)->getInput(output, gdef);
+		if (return_multiple) // If multiple, shove each result into a table
+			lua_newtable(L);
+
+		lua_newtable(L); // items
+		std::vector<ItemStack>::const_iterator iter = input.items.begin();
+		for (u16 j = 1; iter != input.items.end(); iter++, j++) {
+			if (iter->empty())
 				continue;
-			}
-			lua_pushinteger(L,k);
-			lua_pushstring(L,i->name.c_str());
-			lua_settable(L, -3);
+			lua_pushstring(L, iter->name.c_str());
+			lua_rawseti(L, -2, j);
 		}
 		lua_setfield(L, -2, "items");
 		setintfield(L, -1, "width", input.width);
 		switch (input.method) {
 		case CRAFT_METHOD_NORMAL:
-			lua_pushstring(L,"normal");
+			lua_pushstring(L, "normal");
 			break;
 		case CRAFT_METHOD_COOKING:
-			lua_pushstring(L,"cooking");
+			lua_pushstring(L, "cooking");
 			break;
 		case CRAFT_METHOD_FUEL:
-			lua_pushstring(L,"fuel");
+			lua_pushstring(L, "fuel");
 			break;
 		default:
-			lua_pushstring(L,"unknown");
+			lua_pushstring(L, "unknown");
 		}
 		lua_setfield(L, -2, "type");
-	} else {
-		lua_pushnil(L);
-		lua_setfield(L, -2, "items");
-		setintfield(L, -1, "width", 0);
+		lua_pushstring(L, &tmpout.item[0]);
+		lua_setfield(L, -2, "output");
+
+		if (return_multiple)
+			lua_rawseti(L, -2, ++i);
+		else
+			break;
 	}
+
 	return 1;
 }
```

Comment:
I think a cleaner way to design this is with `push_craft_recipe` and `push_craft_recipes` functions, the latter would call the former in a loop and the former could be used directly.